### ISSUE TITLE
"or" kernel applied to slices may be invalid on later operations

### ISSUE THREAD

**Issue by bjchambers** [apache/arrow-rs#1498]
I believe this is due to https://github.com/apache/arrow-rs/issues/807.

Specifically, `or` uses `combine_option_bitmap` which uses `bit_slice` to produce the resulting validity bitmap. But this can lead to cases where the null bitmap has an invalid length (corresponding to the original buffer) while the value array has a length corresponding to the slice.

This in turn can lead to later problems, such as the `nullif` kernel returning an `Err` (internally) when it tries to do a bitwise `&` between the validity bits and the value bits, since they aren't the same length. It then *discards the results of this comparison* (calling `ok()` instead of `unwrap()`) which leads to correct results.

```
(right.values() & &right_bitmap.bits).ok().map(|b| b.not())
```

I believe that either (1+2) or (3) should be done:
1. Replace the `ok()` with `?` to propagate the error.
2. Change `combine_option_bitmap` should to always produce a bitmap of the requested length.
3. Change the logic within the `nullif` kernel to not fail if the right values and right bitmap have different lengths.

Any guidance on which direction the fix should go, and/or ideas on how to implement said fixes?

**Comment by bjchambers** [comment]
Failing test case:

```
    #[test]
    fn test_bool_array_or_sliced() {
        let a = vec![Some(true); 1024];
        let a = BooleanArray::from(a);

        let a = a.slice(0, 2);
        let a: &BooleanArray = a.as_any().downcast_ref().unwrap();

        let b = BooleanArray::from(vec![true, false]);

        let c = or(a, &b).unwrap();

        let expected = BooleanArray::from(vec![Some(true), Some(true)]);

        assert_eq!(c, expected);

        // `nullif` kernel assumes that for a boolean array these are equal
        // since it applies the BitAnd operator between them. Either `logical_or`
        // or `nullif` is broken.
        assert_eq!(
            c.data().null_buffer().unwrap().len(),
            c.data().buffers()[0].len()
        );

        // This is currently silently discarding the condition.
        // Due to the mentioned `ok()`.
        let p = crate::array::Int64Array::from(vec![5, 6]);
        crate::compute::nullif(&p, &c).unwrap();
    }
```

**Comment by jhorstmann** [comment]
Maybe a quick fix could be to use `buffer_bin_and` instead of the `&` operator in `nullif`, which allows passing in the array length.

**Comment by bjchambers** [comment]
It looks like that works, and fixes the behavior of `nullif`. Is the other behavior (the length of the null buffer differing from the length of the boolean data buffer) an error in general?

**Comment by jhorstmann** [comment]
> Is the other behavior (the length of the null buffer differing from the length of the boolean data buffer) an error in general?

I don't think it is, as long as both contain enough space for the length of the array where they are used. In the case of primitive arrays, the length of the validity buffer (in bits) would usually also be larger than the number of elements in the data buffer because it gets rounded up to full bytes.
